dds/DCPS/DiscoveryBase.h

```diff
@@ -1243,14 +1229,61 @@ namespace OpenDDS {
+                                                  const RepoId& remote_id,
+                                                  bool is_discovery_protected)
+      {
+        md.rpc_sequence_number = ++type_lookup_service_sequence_number_;
+        MatchingDataIter md_it = matching_data_buffer_.find(mp);
+        if (md_it != matching_data_buffer_.end()) {
+          md_it->second = md;
+        } else {
+          matching_data_buffer_.insert(std::make_pair(mp, md));
+        }
+        // Store an entry for the first request
+        orig_seq_numbers_.insert(std::make_pair(md.rpc_sequence_number,
+                                                std::make_pair(type_info->minimal.typeid_with_size.type_id,
+                                                               md.rpc_sequence_number)));
+
+        XTypes::TypeIdentifierSeq type_ids;
+        if (type_info->minimal.dependent_typeid_count == -1 ||
+            type_info->minimal.dependent_typeids.length() < (CORBA::ULong)type_info->minimal.dependent_typeid_count) {
+          type_ids.append(type_info->minimal.typeid_with_size.type_id);
+
+          // Get TypeObject of the topic type
+          send_type_lookup_request(type_ids, remote_id, is_discovery_protected, true);
+
+          // Store an entry for the next request
+          orig_seq_numbers_.insert(std::make_pair(++type_lookup_service_sequence_number_,
+                                                  std::make_pair(type_info->minimal.typeid_with_size.type_id,
+                                                                 md.rpc_sequence_number)));
+
+          // Get the dependent TypeIdentifiers of the topic type
+          send_type_lookup_request(type_ids, remote_id, is_discovery_protected, false);
+        } else {
+          type_ids.length(type_info->minimal.dependent_typeid_count + 1);
+          type_ids[0] = type_info->minimal.typeid_with_size.type_id;
+          for (size_t i = 1; i <= (size_t)type_info->minimal.dependent_typeid_count; ++i) {
+            type_ids[i] = type_info->minimal.dependent_typeids[i].type_id;
+          }
+          has_all_dependencies_ = true;
+
+          // Get TypeObjects of the topic type and all of its dependent types
+          send_type_lookup_request(type_ids, remote_id, is_discovery_protected, true);
+        }
+        type_lookup_reply_deadline_processor_->schedule(max_type_lookup_service_reply_period_);
+      }
+
       void
       remove_expired_endpoints(const MonotonicTimePoint& /*now*/)
       {
         ACE_GUARD(ACE_Thread_Mutex, g, lock_);
+        MonotonicTimePoint now = MonotonicTimePoint::now();
         MatchingDataIter end_iter = matching_data_buffer_.end();
 
         for (MatchingDataIter iter = matching_data_buffer_.begin(); iter != end_iter; ) {
-          if (MonotonicTimePoint::now() - iter->second.time_added_to_map >= max_type_lookup_service_reply_period_) {
+          if (now - iter->second.time_added_to_map >= max_type_lookup_service_reply_period_) {
```

Comment:
Do we need to add cleanup for `orig_seq_numbers_`?

Reply:
Yes, it is handled. I'll also add the cleanup for it when replies expire.